internal/provider/data_source_machine.go

```diff
@@ -2,59 +2,126 @@ package provider
 
 import (
 	"context"
+	"fmt"
 
-	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
-	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-framework/datasource"
+	"github.com/hashicorp/terraform-plugin-framework/datasource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/hashicorp/terraform-plugin-log/tflog"
 
 	"github.com/juju/terraform-provider-juju/internal/juju"
 )
 
-func dataSourceMachine() *schema.Resource {
-	return &schema.Resource{
+// Ensure provider defined types fully satisfy framework interfaces.
+var _ datasource.DataSource = &machineDataSource{}
+
+func NewMachineDataSource() datasource.DataSourceWithConfigure {
+	return &machineDataSource{}
+}
+
+type machineDataSource struct {
+	client *juju.Client
+
+	// context for the logging subsystem.
+	subCtx context.Context
+}
+
+type machineDataSourceModel struct {
+	Model     types.String `tfsdk:"model"`
+	MachineID types.String `tfsdk:"machine_id"`
+}
+
+// Metadata returns the full data source name as used in terraform plans.
+func (d *machineDataSource) Metadata(_ context.Context, req datasource.MetadataRequest, resp *datasource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_machine"
+}
+
+func (d *machineDataSource) Schema(_ context.Context, _ datasource.SchemaRequest, resp *datasource.SchemaResponse) {
+	resp.Schema = schema.Schema{
 		Description: "A data source representing a Juju Machine.",
-		ReadContext: dataSourceMachineRead,
-		Schema: map[string]*schema.Schema{
-			"model": {
+		Attributes: map[string]schema.Attribute{
+			"model": schema.StringAttribute{
 				Description: "The name of the model.",
-				Type:        schema.TypeString,
 				Required:    true,
 			},
-			"machine_id": {
+			"machine_id": schema.StringAttribute{
 				Description: "The Juju id of the machine.",
-				Type:        schema.TypeString,
 				Required:    true,
 			},
 		},
```

Comment:
Yeah not sure why don't we have the ID here:

```go
			// ID required by the testing framework
			"id": schema.StringAttribute{
				Computed: true,
			},
```

Reply:
because I used the wrong branch. :-)